ee/fe/modules/500-basic-auth/hooks/generate_passwords.go

```diff
@@ -6,40 +6,115 @@ Licensed under the Deckhouse Platform Enterprise Edition (EE) license. See https
+const locationsInternalKey = "basicAuth.internal.locations"
+const generatedPasswdLength = 20
+
 var _ = sdk.RegisterFunc(&go_hook.HookConfig{
+	Kubernetes: []go_hook.KubernetesConfig{
+		{
+			Name:       secretBinding,
+			ApiVersion: "v1",
+			Kind:       "Secret",
+			NameSelector: &types.NameSelector{
+				MatchNames: []string{secretName},
+			},
+			NamespaceSelector: &types.NamespaceSelector{
+				NameSelector: &types.NameSelector{
+					MatchNames: []string{secretNS},
+				},
+			},
+			// Synchronization is redundant because of OnBeforeHelm.
+			ExecuteHookOnSynchronization: go_hook.Bool(false),
+			ExecuteHookOnEvents:          go_hook.Bool(false),
+			FilterFunc:                   filterHtpasswdSecret,
+		},
+	},
+
 	OnBeforeHelm: &go_hook.OrderedConfig{Order: 10},
 }, generatePassword)
 
-const defaultLocationTemplate = `
-[ {
-  "users": {"admin": "%s"},
-  "location": "/"
-} ]
-`
+func filterHtpasswdSecret(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	secret := &v1.Secret{}
+	err := sdk.FromUnstructured(obj, secret)
+	if err != nil {
+		return nil, fmt.Errorf("cannot convert secret to struct: %v", err)
+	}
+
+	return string(secret.Data["htpasswd"]), nil
+}
+
+const defaultUserName = `admin`
+
+func defaultLocationValues(passwd string) []map[string]interface{} {
+	return []map[string]interface{}{
+		{
+			"users": map[string]interface{}{
+				defaultUserName: passwd,
+			},
+			"location": "/",
+		},
+	}
+}
+
+// Regex to get password from basic auth plain format.
+// Format is: username:{PLAIN}password
+// htpasswd field contains several passwords, so use m flag for multi-line mode.
+var defaultPasswordRe = regexp.MustCompile(`(?m)^\s*` + defaultUserName + `:{PLAIN}(\S+)$`)
 
 func generatePassword(input *go_hook.HookInput) error {
-	_, ok := input.Values.GetOk("basicAuth.locations")
+	// Set values from user controlled configuration.
+	userLocations, ok := input.ConfigValues.GetOk(locationsKey)
 	if ok {
+		input.Values.Set(locationsInternalKey, userLocations.Value())
 		return nil
 	}
 
-	rawLocations := make([]map[string]interface{}, 0)
+	_, ok = input.Values.GetOk(locationsInternalKey)
+	if ok {
+		// No config values, but internal values are set. It's OK, just return.
+		return nil
+	}
 
-	locations := fmt.Sprintf(defaultLocationTemplate, pwgen.AlphaNum(20))
-	err := json.Unmarshal([]byte(locations), &rawLocations)
-	if err != nil {
-		return err
+	// No values, no secret. Module is enabled for the first time, so
+	// generate a new password and prepare a default location.
+	if len(input.Snapshots[secretBinding]) == 0 {
+		locations := defaultLocationValues(pwgen.AlphaNum(generatedPasswdLength))
+		input.Values.Set(locationsInternalKey, locations)
+		return nil
+	}
+
+	// No values, but Secret is present. This can occur when module is enabled
+	// and Deckhouse is restarted later. Restore generated password from the Secret
+	// assuming it is in the first line of htpasswd field.
```

Comment:
```
	// assuming it is in the first line of htpasswd field.
```
What if not?

Reply:
Password restoring is rewritten, no regex, more descriptive message in logs.